### ISSUE TITLE
[11.0] stuck at running unittests when calling write on new env

### ISSUE THREAD

**Issue by oerp-odoo** [odoo/odoo#31878]
Impacted versions:
11.0
Steps to reproduce (sample code, should be the same on any model):
```python
from odoo import models, api
from odoo import registry as registry_get


class ExchangeAction(models.Model):
    """Extend to integrate with mail.thread mixin."""

    _name = 'exchange.action'
    _inherit = ['exchange.action', 'mail.thread']

    def _msg_post(self, e):
        self.ensure_one()
        registry = registry_get(self.env.cr.dbname)
        with registry.cursor() as cr:
            env = api.Environment(cr, self.env.uid, self.env.context)
            # exchange_action = env['exchange.action'].browse(self.id)
            # exchange_action.message_post(body="%s" % e, message_type='comment')
            self.with_env(env).message_post(
                body="%s" % e, message_type='comment')
            cr.commit()

    @api.multi
    def _action_run(self, data=False, fname=False):
        """Override to log exception in related record chatter."""
        try:
            super(ExchangeAction, self)._action_run(data=data, fname=fname)
        except Exception as e:
            self._msg_post(e)
            raise
```
Current behavior:
When running unittests, that involves calling `_msg_post`, it will make those unittests stuck, never to finish and Odoo to be stuck (unless restarted). There are no errors, nothing.

Expected behavior:
It should properly handle operation, such as write, if it is triggered on new environment.

This piece of code works fine, when it is executed using Odoo normally (not via unittests), but for some reason unittests fail to handle it.

I tried two cases: one with using Odoo `SavepointCase` class, another when using `unittest.TestCase` class (on this case, would initiate environment on setUpClass and clean up on `tearDownClass`). In both cases, result is the same. If tests do not involve posting message using new env, then it passes fine, but if it does involve it, it will get stuck.

Did some digging and it turns out it is stuck inside `_write` method, when running `execute` method of cursor -> `odoo/odoo/models.py:3213`. Have no idea what happens inside `execute`.

P.S. my unittest sample (both cases I tried, both get stuck running at `test_exchange_action_log_2`) . Can ignore sftp setup/teardown, which is not relevant with this question, but still posting with it to provide full test case:
```python
import unittest

from odoo.tests.common import SavepointCase, get_db_name
from odoo import registry, api, SUPERUSER_ID
from odoo.exceptions import UserError

from odoo.addons.exchange.tests import common, test_sftp


class TestExchangeLog(unittest.TestCase):
    """Class for exchange.action records logging tests."""

    @classmethod
    def setUpClass(cls):
        """Set up data for tests."""
        super(TestExchangeLog, cls).setUpClass()
        cls.cr = registry(get_db_name()).cursor()
        cls.env = api.Environment(cls.cr, SUPERUSER_ID, {})
        cls.ExchangeAction = cls.env['exchange.action']
        # Use cls.messages to keep track of already created messages,
        # that needs to be cleaned up.
        cls.MailMessage = cls.messages = cls.env['mail.message']
        path_rsa_key = common.TestExchangeCommon.path_rsa_key
        cls.exchange_conn_sftp = cls.env.ref(
            'exchange.exchange_connection_sftp')
        cls.exchange_act_sftp_export = cls.env.ref(
            'exchange.exchange_action_sftp_1')
        cls.process_server = common.setup_server(
            cls,
            test_sftp._setup_sftp,
            args=(path_rsa_key,),
            kwargs={'port': test_sftp.PORT, 'delay': 0.01})
        cls.transfer_obj = common.connect_to_server(
            cls,
            cls.exchange_conn_sftp,
            pre_handler=test_sftp._attach_pkey,
            pre_args=[path_rsa_key])

    def _get_messages(self, exchange_act_id):
        return self.MailMessage.search([
            ('model', '=', 'exchange.action'),
            ('res_id', '=', exchange_act_id)
        ])

    def test_exchange_action_log_1(self):
        """Run action normally, to not get any exceptions.

        No log messages should be created.
        """
        self.ExchangeAction._cron_action_run_multi(
            domain=[('id', '=', self.exchange_act_sftp_export.id)])
        self.assertFalse(self._get_messages(self.exchange_act_sftp_export.id))

    def test_exchange_action_log_2(self):
        """Run action to get exception.

        Log message should be created when _cron_action_run_multi is
        called, but not when _action_run is called directly.
        """
        exchange_act_id = self.exchange_act_sftp_export.id
        self.exchange_act_sftp_export.path = 'not_existing_path'
        with self.assertRaises(UserError):
            self.ExchangeAction._cron_action_run_multi(
                domain=[('id', '=', exchange_act_id)])
        messages = self._get_messages(exchange_act_id)
        self.assertEqual(len(messages), 1)

        with self.assertRaises(UserError):
            self.exchange_act_sftp_export._action_run()
        messages = self._get_messages(exchange_act_id)
        self.assertEqual(len(messages), 2)
        # Add messages for cleanup.
        self.messages |= messages

    def tearDown(self):
        """Clean up after each test."""
        super(TestExchangeLog, self).tearDown()
        # Remove created messages.
        self.messages.unlink()

    @classmethod
    def tearDownClass(cls):
        """Kill SFTP server and its relations."""
        super(TestExchangeLog, cls).tearDownClass()
        common.tear_down_server(
            cls, cls.exchange_conn_sftp, port=test_sftp.PORT)
        cls.cr.close()


class TestExchangeLogOld(SavepointCase):
    """Class for exchange.action records logging tests."""

    @classmethod
    def setUpClass(cls):
        """Set up data for tests."""
        super(TestExchangeLogOld, cls).setUpClass()
        cls.ExchangeAction = cls.env['exchange.action']
        cls.MailMessage = cls.env['mail.message']
        path_rsa_key = common.TestExchangeCommon.path_rsa_key
        cls.exchange_conn_sftp = cls.env.ref(
            'exchange.exchange_connection_sftp')
        cls.exchange_act_sftp_export = cls.env.ref(
            'exchange.exchange_action_sftp_1')
        cls.process_server = common.setup_server(
            cls,
            test_sftp._setup_sftp,
            args=(path_rsa_key,),
            kwargs={'port': test_sftp.PORT, 'delay': 0.01})
        cls.transfer_obj = common.connect_to_server(
            cls,
            cls.exchange_conn_sftp,
            pre_handler=test_sftp._attach_pkey,
            pre_args=[path_rsa_key])

    def _get_messages(self, exchange_act_id):
        return self.MailMessage.search([
            ('model', '=', 'exchange.action'),
            ('res_id', '=', exchange_act_id)
        ])

    def test_exchange_action_log_1(self):
        """Run action normally, to not get any exceptions.

        No log messages should be created.
        """
        self.ExchangeAction._cron_action_run_multi(
            domain=[('id', '=', self.exchange_act_sftp_export.id)])
        self.assertFalse(self._get_messages(self.exchange_act_sftp_export.id))

    def test_exchange_action_log_2(self):
        """Run action to get exception.

        Log message should be created when _cron_action_run_multi is
        called, but not when _action_run is called directly.
        """
        exchange_act_id = self.exchange_act_sftp_export.id
        self.exchange_act_sftp_export.path = 'not_existing_path'
        with self.assertRaises(UserError):
            self.ExchangeAction._cron_action_run_multi(
                domain=[('id', '=', exchange_act_id)])
        messages = self._get_messages(exchange_act_id)
        self.assertEqual(len(messages), 1)
        with self.assertRaises(UserError):
            self.exchange_act_sftp_export._action_run()
        messages = self._get_messages(exchange_act_id)
        self.assertEqual(len(messages), 2)

    @classmethod
    def tearDownClass(cls):
        """Kill SFTP server and its relations."""
        common.tear_down_server(
            cls, cls.exchange_conn_sftp, port=test_sftp.PORT)
        super(TestExchangeLogOld, cls).tearDownClass()
```

This is how log looks like, if I run any of these cases (last lines of log, and it stays that way, never finishing unittests):

```
2019-03-15 12:32:37,511 28825 INFO toc_198 odoo.addons.exchange_log.tests.test_exchange_log: test_exchange_action_log_1 (odoo.addons.exchange_log.tests.test_exchange_log.TestExchangeLog)
2019-03-15 12:32:37,511 28825 INFO toc_198 odoo.addons.exchange_log.tests.test_exchange_log: ` Run action normally, to not get any exceptions.
2019-03-15 12:32:37,524 28825 INFO toc_198 paramiko.transport.sftp: [chan 0] sftp session closed.
2019-03-15 12:32:37,536 28825 INFO toc_198 odoo.addons.exchange_log.tests.test_exchange_log: test_exchange_action_log_2 (odoo.addons.exchange_log.tests.test_exchange_log.TestExchangeLog)
2019-03-15 12:32:37,536 28825 INFO toc_198 odoo.addons.exchange_log.tests.test_exchange_log: ` Run action to get exception.
2019-03-15 12:32:38,512 28825 INFO ? paramiko.transport: Connected (version 2.0, client paramiko_2.4.2)
2019-03-15 12:32:38,512 28845 INFO ? paramiko.transport: Connected (version 2.0, client paramiko_2.4.2)
2019-03-15 12:32:38,524 28845 INFO ? paramiko.transport: Auth granted (publickey).
2019-03-15 12:32:38,524 28825 INFO ? paramiko.transport: Authentication (publickey) successful!
2019-03-15 12:32:38,527 28825 INFO toc_198 paramiko.transport.sftp: [chan 0] Opened sftp connection (server version 3)
```
This is how it looks like, if I ignore `test_exchange_action_log_2` (test that tries to use new env and make some changes in database). Last lines of log after running unittests.:

```
2019-03-15 12:35:08,259 29100 INFO toc_198 odoo.addons.exchange_log.tests.test_exchange_log: test_exchange_action_log_1 (odoo.addons.exchange_log.tests.test_exchange_log.TestExchangeLog)
2019-03-15 12:35:08,259 29100 INFO toc_198 odoo.addons.exchange_log.tests.test_exchange_log: ` Run action normally, to not get any exceptions.
2019-03-15 12:35:08,275 29100 INFO toc_198 paramiko.transport.sftp: [chan 0] sftp session closed.
2019-03-15 12:35:08,287 29100 INFO toc_198 sh.command: <Command '/bin/fuser -k 3373/tcp'>: starting process
2019-03-15 12:35:08,292 29100 INFO toc_198 sh.command: <Command '/bin/fuser -k 3373/tcp', pid 29140>: process started
2019-03-15 12:35:08,384 29100 INFO toc_198 sh.command: <Command '/bin/fuser -k 3373/tcp', pid 29140>: process completed
2019-03-15 12:35:08,386 29100 INFO toc_198 odoo.addons.exchange_log.tests.test_exchange_log: Ran 1 test in 0.187s
2019-03-15 12:35:08,386 29100 INFO toc_198 odoo.addons.exchange_log.tests.test_exchange_log: OK
2019-03-15 12:35:08,436 29100 INFO toc_198 odoo.modules.loading: 25 modules loaded in 0.67s, 27 queries
2019-03-15 12:35:08,556 29100 INFO toc_198 odoo.modules.registry: verifying fields for every extended model
2019-03-15 12:35:08,646 29100 INFO toc_198 odoo.modules.loading: Modules loaded.
2019-03-15 12:35:08,649 29100 INFO toc_198 odoo.service.server: All post-tested in 0.00s, 0 queries
```

**Comment by pedrobaeza** [comment]
You shouldn't do commit in your model, and no need for tearDown.

Any way, the error you are mentioning corresponds to a custom module or code, so the problem should be there in your code. This issue tracker is for notifying problems with standard Odoo code, and thus I'm closing this.

Try to share your question, doubt or problem in the official Odoo forums (https://www.odoo.com/forum/help-1) or on the mailing lists (https://www.odoo.com/groups) to see if someone can help you.

**Comment by oerp-odoo** [comment]
@pedrobaeza apparently it seems there is problem with my code (still have no idea what is wrong there). Sorry that I jumped to conclusions. With simpler code and unittests for it, it seems to work as expected. 

I understand that it is not advised to do manual commit, but sometimes it is needed, as was in my case. because I needed to commit even if exception is raised. If I would do normal transaction (without new env), it would be rolled back after exception is raised and my message would not be created.

And for tearDown, it is needed (in first case I used `TestCase` class which is plain unittest class, that does not know how to clean up odoo stuff (or does it?). Also in second test class case, might also need to do manual cleaning, because apparently Odoo does not cleanup stuff that was modified with new env).
